### ISSUE TITLE
Build a harness that scores the pipeline as actually submitted

### ISSUE THREAD

**Issue by vltnkiz** [vltnkiz/numerai#28]
Part of #24

## Question

Build the measurement backbone: a harness that scores the **pipeline as actually submitted** on validation eras.

Today `score_validation` runs on raw per-model predictions at `pipeline.py:129-146`, *before* combination and *before* neutralization. Nothing in the repo has ever measured the artifact that gets uploaded. Every "is A better than B" question on this map is unanswerable until this exists.

The harness must, over validation eras:

- Apply the **full** submitted transform — combine → neutralize → rank-normalize — not a subset.
- Report per-era correlation, mean, std, Sharpe, and smart Sharpe (reuse `score_validation`), plus feature exposure so neutralization's effect is visible rather than inferred.
- Compare configurations in one run: linear alone, era-boost alone, the blend, and each under varying neutralization target and proportion.
- Be reproducible — fixed seeds, recorded config, results written to a run directory.

Open sub-questions to settle while building: does neutralization run per-era on validation as it does on single-era live data? Does the current 3.85 GB validation set need subsampling to keep iteration tolerable, and does subsampling bias era statistics?

**Answer must record**: where the harness lives, how to invoke it, and the **baseline numbers for today's production linear model** that every later comparison is measured against.

---

## Updated by [#26](https://github.com/vltnkiz/numerai/issues/26) — the harness must score the *paid* metrics

Research on #26 changed what this harness has to compute. Two corrections to the brief above:

1. **`score_validation` does not measure what Numerai pays on.** It uses plain Spearman. The paid metric is `numerai_corr`: rank → gaussianize (`norm.ppf`) → raise both preds and centered target to the power 1.5 → Pearson. Implement it per era. Sharpe and smart Sharpe remain useful risk read-outs but are **not** payout metrics.

2. **MMC is the metric that actually decides the neutralization question, and it is locally computable.** Payout is `0.75·corr20 + 2.25·mmc20` — MMC carries **3× the weight of CORR**. `v5.0/meta_model.parquet` is downloadable (15 MB) and verified: 639,426 rows, column `numerai_meta_model`, zero nulls, covering **validation eras 1133–1228 (96 eras)**. MMC is `covariance(centered_target, orthogonalize(gaussian(rank(preds)), gaussian(rank(meta_model))))`.

   → The harness must report **CORR and MMC**, and the headline comparison figure should be the payout-weighted `0.75·corr + 2.25·mmc`. A config comparison on raw corr alone will pick the wrong architecture.

   → MMC comparisons are restricted to the 96-era meta-model window. Report CORR on the full validation span *and* on that window, so the two are comparable.

3. **`rank_normalize` is free** — `numerai_corr` and MMC are exactly invariant to strictly monotone transforms (verified: delta `0.00e+00` across four transforms). No need to model its effect.

4. **Reproducibility is not a concern** — XGBoost fits here are bit-identical across runs despite `random_state=None` (see [#27](https://github.com/vltnkiz/numerai/issues/27)). Measured margins are not seed noise.

**Caveat on every number this harness produces**: v5.0's `target` is the cyrus-era target, but Numerai's payout target has been **Ender20 since 2026-01-01**, and `target_ender_20` does not exist in v5.0. Numbers here are measured against a target Numerai no longer pays on. This does not block the harness — relative comparisons between configs remain informative — but it caps what they prove. Owned by [#21](https://github.com/vltnkiz/numerai/issues/21).

**Comment by vltnkiz** [comment]
## Answer

The harness exists, is validated, and has produced the baseline. **It also produced a result that reframes the rest of this map: on Numerai's paid metric, every configuration we can currently build is payout-negative.**

### Where it lives, and how to run it

| Piece | Path |
| --- | --- |
| Fit + score stages | `prod/zemir_01/zemir/harness.py` |
| Paid metrics | `prod/zemir_01/zemir/scoring.py` (`era_numerai_corr`, `era_mmc`, `era_max_feature_corr`, `summarize_era_scores`) |
| `ScoringConfig` + sweep | `prod/zemir_01/zemir/config.py` |
| Entrypoints | `scripts/fit_harness.py`, `scripts/score_harness.py` |
| Artifacts | `prod/zemir_01/runs/harness/<run_id>/` |

```bash
python scripts/fit_harness.py --model ensemble           # ~35 min, caches raw validation predictions
python scripts/fit_harness.py --model ensemble --smoke   # seconds, identical code path
python scripts/score_harness.py [run_id]                 # 2m19s, sweeps 15 configs
```

Split at the only expensive seam. Fitting caches every model's **raw** validation predictions once (`validation_predictions.parquet` + `fit_config.json`); scoring sweeps blends and neutralization proportions over that cache and re-runs freely. Anything that changes a *model* invalidates the cache; anything downstream of `.predict()` does not. `score_harness.py` reads the cache's own `fit_config.json`, so a sweep cannot be scored against the wrong dataset.

Neither script contains or imports a submission path, per [#25](https://github.com/vltnkiz/numerai/issues/25).

### The baseline

Fit `20260825T194439Z-ensemble` — `LIVE` config, v5.0 `small` (42 features), 574 train eras, **653 validation eras**, 4,078,596 rows. MMC window: **95 eras, 1133–1227**.

**Today's production model is `linear_p0.5`:**

| | full span (653 eras) | MMC window (95 eras) |
| --- | --- | --- |
| `numerai_corr` | **0.012153** | **0.002048** |
| sharpe / smart sharpe | 0.695294 / 0.287474 | — |
| MMC | — | **−0.003539** (sharpe −0.259721) |
| max feature exposure | 0.297398 | |
| **payout** `0.75·corr + 2.25·mmc` | | **−0.006427** |

Full sweep in `runs/harness/20260825T194439Z-ensemble/scores.csv`, ranked by payout:

| config | mean_corr (653) | sharpe | corr (95) | mean_mmc | max_feat_corr | payout |
| --- | --- | --- | --- | --- | --- | --- |
| linear_p1 | −0.000780 | −0.063 | −0.000488 | −0.000648 | 0.004842 | −0.001823 |
| linear+era_boost_p1 | 0.009207 | 0.577 | 0.001673 | −0.003062 | 0.021223 | −0.005636 |
| **linear_p0.75 / p0.5 / p0.25 / p0** | **0.012153** | **0.695** | **0.002048** | **−0.003539** | **0.297398** | **−0.006427** |
| era_boost_p1 | 0.011667 | 0.670 | 0.001306 | −0.003838 | 0.019005 | −0.007657 |
| linear+era_boost_p0 | 0.015035 | 0.831 | 0.002404 | −0.004368 | 0.282715 | −0.008024 |
| era_boost_p0.75 | 0.014678 | 0.808 | 0.001630 | −0.004468 | 0.096094 | −0.008831 |
| era_boost_p0 | 0.016867 | 0.899 | 0.002490 | −0.004827 | 0.251140 | −0.008994 |
| era_boost_p0.25 | **0.016914** | **0.900** | 0.002364 | −0.004863 | 0.219474 | −0.009170 |
| era_boost_p0.5 | 0.016334 | 0.876 | 0.002028 | −0.004788 | 0.169769 | −0.009251 |
| linear+era_boost_p0.75 | 0.014447 | 0.821 | 0.001202 | −0.004843 | 0.163306 | −0.009996 |
| linear+era_boost_p0.25 | 0.015716 | 0.838 | 0.001427 | −0.004998 | 0.269528 | −0.010176 |
| linear+era_boost_p0.5 | 0.015655 | 0.834 | 0.000533 | −0.005475 | 0.237418 | −0.011919 |

### 🔴 Every configuration is payout-negative, and the CORR ranking is inverted against it

Best CORR is `era_boost_p0.25` (0.016914). Best *payout* is `linear_p1` (−0.001823) — the **degenerate** config, whose signal neutralization has zeroed. It ranks first only because a near-noise prediction detracts least from the meta model. It is not a candidate; it is a warning that the payout ordering is currently dominated by "how little you hurt the crowd", not by skill.

This is exactly the failure mode [#26](https://github.com/vltnkiz/numerai/issues/26) predicted for comparing on raw CORR — but sharper than expected: MMC is not merely decisive, it is negative everywhere, so the two metrics rank the sweep in almost opposite order.

### The harness is validated, so the result stands

Three independent checks:

1. **MMC self-test.** The meta model's MMC against itself is `8.94e-19`, max `4.18e-17` per era — exactly zero to float precision, as the definition requires.
2. **Negative MMC is arithmetic, not a bug.** Over the same 95 eras the meta model's own `numerai_corr` is **0.022846** against our 0.002048 (linear) / 0.002490 (era_boost) — the crowd is ~11× stronger — while our predictions are 0.27 / 0.35 rank-correlated with it. A weaker prediction pointing the same way as the crowd has negative MMC by construction.
3. **It reproduces the map's established fact 1 independently.** `linear_p0`, `p0.25`, `p0.5`, `p0.75` come out identical to six decimals on CORR, MMC **and** feature exposure (0.012153 / −0.003539 / 0.297398). Only `p1` differs. Neutralizing a linear model against its own features is exactly a scalar shrink, confirmed at full scale on the paid metrics.

### 🔴 Signal decays hard, and era boosting's advantage decays with it

Mean `numerai_corr` by era octile, un-neutralized:

| eras | linear | era_boost | blend |
| --- | --- | --- | --- |
| 575–656 | +0.02453 | +0.02789 | +0.02750 |
| 657–738 | +0.01075 | +0.02209 | +0.01627 |
| 739–819 | +0.01476 | +0.02144 | +0.01913 |
| 820–901 | +0.00450 | +0.01360 | +0.00955 |
| 902–982 | +0.01727 | +0.02310 | +0.02052 |
| 983–1064 | +0.01285 | +0.01579 | +0.01499 |
| 1065–1145 | +0.01074 | +0.01027 | +0.01083 |
| **1146–1227** | **+0.00189** | **+0.00079** | **+0.00156** |

Two things follow, and both bear on [#30](https://github.com/vltnkiz/numerai/issues/30) and [#31](https://github.com/vltnkiz/numerai/issues/31):

- **The full-span mean flatters us by an order of magnitude.** 0.0122–0.0169 across 653 eras becomes 0.0005–0.0025 in the last 95 — the only eras where MMC is measurable, and the ones closest to the live regime. Any comparison quoted on the full span is a comparison on a market that no longer exists.
- **Era boosting's edge is entirely historical.** It beats linear by 40% full-span (0.016867 vs 0.012153), ties it at eras 1065–1145, and *loses* to it in the most recent octile (+0.00079 vs +0.00189). The map's fog entry "what if era boosting is actively harmful" now has evidence pointing at *recently*, yes.

### Sub-questions from the brief

- **Does neutralization run per era on validation?** Yes — `era_feature_projection` groups by era, mirroring `neutralize_predictions`, which production applies to the single live era.
- **Does validation need subsampling?** No. 15 configs × 653 eras × 4.08M rows scores in **2m19s**. Era statistics stay unbiased because nothing is dropped.

### Two implementation notes worth carrying forward

1. **A production/harness divergence, found and closed.** `combine_predictions` was only ever called for multi-model runs; `run_pipeline` special-cased a lone model to its raw prediction. The harness called it unconditionally, which per-era *ranked* the linear model before neutralizing — and since ranking is non-linear, that silently broke the scalar-shrink identity. The single-model rule now lives inside `combine_predictions` so the two cannot disagree.
2. **Neutralization is affine in proportion:** `scores − p · projection`, and the projection is independent of `p`. Sweeping proportions therefore costs one least-squares solve per era, not one per config — the difference between 2 minutes and ~3 hours. `neutralize` was rewritten in terms of the same `project` helper so there is one definition of the math (verified bit-identical, max diff `0.0`).

### Caveats on every number above

- **Wrong target.** v5.0's `target` is cyrus-era; Numerai has paid on Ender20 since 2026-01-01 ([#26](https://github.com/vltnkiz/numerai/issues/26), owned by [#21](https://github.com/vltnkiz/numerai/issues/21)). These numbers measure a target nobody is paid on.
- **Narrow features.** 42 features (`small`) against a crowd using far more. That is a plausible cause of both the weak CORR and the negative MMC — and feature-set width is [#21](https://github.com/vltnkiz/numerai/issues/21)'s, not this map's.
- **MMC covers 95 of 653 eras.** Relative comparisons between configs remain informative; absolute payout numbers do not transfer to the live regime unchanged.

**Comment by vltnkiz** [comment]
Resolved — see resolution comment above. Harness lives in `zemir/harness.py` + `scripts/fit_harness.py` / `scripts/score_harness.py`; baseline for today's production `linear_p0.5` is corr 0.012153 (653 eras) / 0.002048 + MMC −0.003539 (95-era MMC window) → payout **−0.006427**. Every one of the 15 configurations is payout-negative; surfaced [#34](https://github.com/vltnkiz/numerai/issues/34).
